doc/compiler/ObjectLockword.md

```diff
@@ -0,0 +1,386 @@
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Background
+To facilitate synchronization, Java objects have a lockword to track information regarding locking. In the simple case, the per object lockword tracks which thread has locked the object and other threads need to wait for the object to be unlocked before they themselves can lock the object and continue into their own critical section. It is possible for the same thread to lock the same object multiple times before unlocking the object. This is known as nested locking. For an object to be fully unlocked so it can be locked by other threads the object must be unlocked an equal number of times to the number of times it was locked and the lockword also needs to keep track of this.
+
+Since multiple threads may have an interest in updating the lockword to lock the object it is frequently necessary to use CAS to update the lockword. This is slower than using plain loads and stores to update the lockword. As a result, there is an interest in using lock reservation optimization to potentially speed things up. Lock reservation allows threads to reserve an object. What this means is that if a thread that owns the reservation on an object attempts to lock the object, it will be able to use plain loads and stores to update the lockword. The tradeoff is that if a different thread attempts to lock the reserved object this will trigger a very expensive process to cancel the lock reservation.
+
+# Original Lockword format
+## Original Lockword Bit Pattern
+```
+64 bit lockword:
+| TID (56 bits) | RC (5 bits) | Res | FLC | Infl |
+32 bit lockword:
+| TID (24 bits) | RC (5 bits) | Res | FLC | Infl |
+
+ TID = Thread ID excluding lowest 8 bits
+  RC = Recursion Count
+ Res = Reserved Bit
+ FLC = Flatlock Contention Bit
+Infl = Inflated Bit
+```
+
+The lockword format is largely defined in `runtime/oti/j9nonbuilder.h`. Defines are used to indicate which bit position has what meaning and to provide some useful bit masks.
+
+The TID field takes advantage of how the lowest 8 bits in a Thread ID are always 0 so they don't need to be stored.
+
+The RC or Recursion Count field is primarily used to track nested locking. One key point is the RC field is slightly different for reserved locks and unreserved locks. For reserved locks the RC field is equal to the nesting depth of the locks. This means that if the object was locked 3 times before being unlocked then the RC field will have a value of 3. If the object is not locked at all, the RC field will be 0. For unreserved locks the RC field is one less than the nesting depth. This means that if the object was locked 3 times without being unlocked then the RC field will have a value of 2. If the object is not locked at all, the RC field will be 0 but the TID field must also be 0 for the object to be considered fully unlocked.
```

Comment:
Would be helpful to note the behaviour here on overflow of the counter...

Reply:
I will add that in.